src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit.cpp

```diff
@@ -1196,106 +1241,185 @@ Array<Pair<Index, Pass>> Orbit::ComputePasses(const Array<State>& aStateArray, c
         }
     }
 
-    const model::Tabulated tabulated = model::Tabulated(aStateArray, Interpolator::Type::BarycentricRational);
+    const orbit::model::Tabulated tabulated =
+        orbit::model::Tabulated(aStateArray, anInitialRevolutionNumber, Interpolator::Type::BarycentricRational);
+
+    const Interval interval =
+        Interval::Closed(aStateArray.accessFirst().accessInstant(), aStateArray.accessLast().accessInstant());
+
+    const Array<Pass> passes = Orbit::ComputePassesWithModel(tabulated, interval);
+
+    Array<Pair<Index, Pass>> passMap;
+    passMap.reserve(passes.getSize());
+
+    Index stateIndex = 0;
+    Index passIndex = 0;
+
+    Pass const* currentPassPtr = &passes[passIndex];
```

Comment:
Why would you use a shared or unique pointer here? We are not allocating memory here we are merely just getting a pointer to the element.

Reply:
Ah ok good point, I thought you could just simply create a shared pointer to an object that already exists more simply, but it turns out you have to use a custom deleter or something like that which complicates things.

However, my point still stands, why use raw pointers unless absolutely necessary? I think we should avoid them unless we have a good reason to use them. In this case, I get that you're trying to not copy the pass at `passes[passIndex]` to a `currentPass` variable at each iteration of the loop, but I don't think this needs a pointer, why not just use the  `passes[passIndex]` itself like this:
```cpp
while (stateIndex < aStateArray.getSize())
{
    const Instant& stateInstant = aStateArray[stateIndex].accessInstant();

    if (passes[passIndex]->accessInstantAtPassBreak().isDefined() &&
        stateInstant >= passes[passIndex]->accessInstantAtPassBreak())
    {
        passMap.add({stateIndex, passes[passIndex]});
        ++passIndex;
    }

    stateIndex++;
}
```
Also, I feel like with all optimizations turned on the compiler probably optimizes out the "copying" to the currentPass variable on each iteration.

Reply:
Your solution would work fine, but it would result in copying the pass object, it also results in repeated indexing of the passes array, which is fine cause it's constant complexity.
However, I don't believe we should live in fear of pointers. Pointers like any other programming paradigm have their place. You can segfault your program and memory leak it without using raw pointers. You can segfault it with Smart pointers as well.
Instead I would suggest understanding how they work and when they are effective. We are utilizing C++ for it's ability to do precise memory management, in this case it's relatively unnoticeable as we're operating on 100s of passes at most, but if this pattern was copied for something else that is 1000s or more we would start to see the impact.
I'll update the code but it's food for thought.

Reply:
To me the cost benefit of trying to do precise memory management at that granular of a level is not worth it, especially with the impact it can have on dev velocity. To me, the reason we use C++ is for code performance, that's it, I wouldn't say that about precise memory management. Using C++ for OSTk impacts us negatively in virtually every other way, but we stick with it because it is so damn fast and we need it to be that fast.

Sure you can segfault without pointers, but screwing something up with pointers is easy to do if you don't 100% know what you're doing.

Also can't you do it with a non-const reference instead of a raw pointer?
```cpp 
Pass& currentPass = passes[passIndex];

while (stateIndex < aStateArray.getSize())
{
    const Instant& stateInstant = aStateArray[stateIndex].accessInstant();

    if (currentPass.accessInstantAtPassBreak().isDefined() &&
        stateInstant >= currentPass.accessInstantAtPassBreak())
    {
        passMap.add({stateIndex, currentPass});
        ++passIndex;
		currentPass = passes[passIndex]
    }

    stateIndex++;
}
```

Also also, are you sure the Pass object is not being copied at this step anyways? `passMap.add({stateIndex, *currentPassPtr});`

Reply:
> the reason we use C++ is for code performance
This doesn't come for free. You can write code that is not performant in C++. OSTk for example is atleast 1-3 orders of magnitude slower than propagators I've used in the past.

> Using C++ for OSTk impacts us negatively in virtually every other way, but we stick with it because it is so damn fast and we need it to be that fast.

? I'm not quite sure I understand what you mean here.

> Also also, are you sure the Pass object is not being copied at this step anyways? passMap.add({stateIndex, *currentPassPtr});
It's being copied there once as it has to be added to the map. It's not making an additional copy of the `currentPass`.

> I wouldn't say that about precise memory management
The hallmark of modern performance optimization **is** memory management.

You can have the exact same issues you have with a non-const reference that you can with a pointer. Why is that any "safer"? If you make a reference to something that goes out of scope you can segfault.
Either way. It's been amended.